File: C/Searching/binary_search.c
#include<stdio.h>
#include<stdlib.h>


void bubbleSort(int ptr[], int n)
{
    int i, j, flag = 0, temp;
    for (i=0;i<n-1;i++)
    {
        for (j=0;j<n-1-i;j++)
        {
            if (ptr[j] > ptr[j+1])
            {
                flag = 1;
                temp = ptr[j];
                ptr[j] = ptr[j+1];
                ptr[j+1] = temp;
            }
        }
        
        if (flag == 0)
        {
            printf("\nArray is sorted.");
            return;
        }
            
    }
    printf("[");
     for (i=0;i<n;i++)
    {
        printf("%d, ", ptr[i]);
    }
    printf("]");
  

}

void binary_search(int A[], int n, int val)
{
    int loc, beg, end, mid, flag = 0;
    loc = 0;
    beg = 0;
    end = n-1;


    while ((beg <= end) && (val != A[mid]))
    {
        mid = (beg+end)/2;
        if (val == A[mid])
        {
            printf("\nSearch found at %d", mid+1);
            loc = mid;
            flag = flag + 1;
        }

        else if (val < A[mid])
        {
            end = mid - 1;
        }

        else {
            beg = mid + 1;
        }
    }
    if (flag == 0)
    {
        printf("\nSearch not found.");
    }
    
}


int main()
{
    int n, i, query;
    int *ptr;
    printf("\nEnter how many elements?");
    scanf("%d", &n);
    ptr = (int*)malloc(n*sizeof(int));
    for (i = 0;i<n;i++)
    {
        scanf("%d", &ptr[i]);

    }
    printf("\nThe data is ");
    for (i = 0;i<n;i++)
    {
       printf("\n\t%d", ptr[i]);
    }
    printf("\nEnter the search value:");
    scanf("%d", &query);
    bubbleSort(ptr,n);
    binary_search(ptr, n, query);

   
    return 0;

}

File: Python/Searching/binary_search.py
import random
import time
import math

class Data:
    def __init__(self):
        self.items = []
        self.num = None
        self.temp = 0
        self.swapped = True

    def get(self):
        self.num = int(input("Enter how many number:"))
        for i in range(self.num):
            item = int(input("Enter data:"))
            self.insert(item)


    def insert(self, item):
        self.items.append(item)

   

    def random(self):
        self.num = int(input("Enter how many number:"))
        item = [x+1 for x in range(self.num)]
        for i in item:
            self.items.append(i)
        random.seed(time.time())
        random.shuffle(self.items)

    def display(self):
        print(f'{self.items}')

       

    def swap(self, items, i, j):
        if i != j:
            items[i], items[j] = items[j], items[i]



    def sort(self, items):
        if len(items) == 1:
            print(items)
            return           # Exit from function

        self.swapped = True
        for i in range(len(items)-1):
            if not self.swapped:
                break
            self.swapped = False
            for j in range(len(items)-1-i):
                if (items[j] > items[j+1]):
                    self.swap(items, j, j+1)
                    self.swapped = True
                print(items)


    def search(self):
        self.sort(self.items)
        val = int(input("Enter the data you want to search:"))
        beg = 0
        end = len(self.items) -1
        pos = -1

        while(beg <= end):
            mid = (beg+end)/ 2
            mid = math.ceil(mid)
            if (val == self.items[mid]):
                pos = mid
                print(f'{val} is present at {pos+1}')
                break

            elif (self.items[mid] > val):
                end = mid - 1

            else:
                beg = mid + 1


        if (pos == -1):
            print("Search not found")
        

            





    

if __name__ == "__main__":    
    data = Data()
   # data.get()
    data.random()
    data.display()
    data.search()
    
        
